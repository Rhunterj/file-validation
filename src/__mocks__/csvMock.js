export const csvMock = `Reference,Account Number,Description,Start Balance,Mutation,End Balance
123,NL123,description,123,+123,123
`;
