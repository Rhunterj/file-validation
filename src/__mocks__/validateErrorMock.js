export const validateErrorMock = {
  date: "Sun Jan 01 2017",
  time: "01:00:00",
  duplicateKeys: [
    {
      accountNumber: "NL90ABNA0585647886",
      description: "Flowers for Richard Bakker",
      endBalance: 109.53,
      mutation: 14.63,
      reference: 138932,
      startBalance: 94.9,
    },
    {
      accountNumber: "NL93ABNA0585619023",
      description: "Candy from Vincent de Vries",
      endBalance: 6368,
      mutation: -939,
      reference: 138932,
      startBalance: 5429,
    },
  ],
  invalidEndBalance: [
    {
      accountNumber: "NL93ABNA0585619023",
      description: "Candy from Vincent de Vries",
      endBalance: 6368,
      mutation: -939,
      reference: 138932,
      startBalance: 5429,
    },
  ],
  itemsChecked: 5,
};
