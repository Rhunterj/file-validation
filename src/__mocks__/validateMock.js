export const validateMock = [
  {
    reference: 138932,
    accountNumber: "NL90ABNA0585647886",
    description: "Flowers for Richard Bakker",
    startBalance: 94.9,
    mutation: 14.63,
    endBalance: 109.53,
  },
  {
    reference: 138932,
    accountNumber: "NL93ABNA0585619023",
    description: "Candy from Vincent de Vries",
    startBalance: 5429,
    mutation: -939,
    endBalance: 6368,
  },
  {
    reference: 101339,
    accountNumber: "NL27SNSB0917829871",
    description: "Tickets from Rik King",
    startBalance: 84.46,
    mutation: 40.45,
    endBalance: 124.91,
  },
  {
    reference: 119582,
    accountNumber: "NL32RABO0195610843",
    description: "Subscription from Vincent de Vries",
    startBalance: 38.86,
    mutation: 28.77,
    endBalance: 67.63,
  },
  {
    reference: 180148,
    accountNumber: "NL32RABO0195610843",
    description: "Candy for Rik de Vries",
    startBalance: 51.01,
    mutation: -25.59,
    endBalance: 25.42,
  },
];
