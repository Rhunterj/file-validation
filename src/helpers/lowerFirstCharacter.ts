export const lowerFirstCharacter = (str: string) => {
  return str[0].toLowerCase() + str.slice(1);
}