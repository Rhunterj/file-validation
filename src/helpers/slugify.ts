export const slugify = (str: string): string => {
  if (!str) {
    return str
  }

  return str.toLowerCase().replace(/[^a-zA-Z0-9]+(.)/g, (m, chr) => chr.toUpperCase());
}