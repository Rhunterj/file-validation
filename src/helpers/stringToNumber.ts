export const convertStringToNumber = (str: string) => {
  const number = Number(str);

  if (isNaN(number)) {
    return str;
  }

  return number;
}