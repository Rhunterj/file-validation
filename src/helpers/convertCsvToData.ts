import { convertedDataType } from "../types";
import { slugify } from "./slugify";
import { convertStringToNumber } from "./stringToNumber";

export const convertCsvToData = (csv: string): convertedDataType[] => {
  const lines = csv.split("\n").slice(1);
  lines.pop(); // remove last empty line

  return lines.map(line => {
    const values = line.split(",");
    const headers = csv.split("\n")[0].split(",");

    return headers.reduce((obj: any, header: string, index: number) => {
      obj[slugify(header)] = convertStringToNumber(values[index]);
      return obj;
    }, {}) as convertedDataType
  });
}