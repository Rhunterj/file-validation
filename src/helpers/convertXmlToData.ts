import { convertedDataType } from "../types";
import { convertStringToNumber } from "./stringToNumber";

export const convertXmlToData = (xml: string): convertedDataType[] => {
  const parser = new DOMParser();
  const xmlDoc = parser.parseFromString(xml, "text/xml");
  const records = xmlDoc.querySelectorAll('record');

  return Array.from(records, record => {
    const obj: Record<string, any> = {};
    obj['reference'] = Number(record.getAttribute("reference"));
    Array.from(record.children).forEach(child => {
      obj[child.tagName] = convertStringToNumber(child.innerHTML);
    });
    return obj as convertedDataType;
  });
}