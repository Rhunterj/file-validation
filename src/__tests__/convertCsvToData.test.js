import { convertCsvToData } from '../helpers/convertCsvToData';
import { csvMock } from '../__mocks__/csvMock';

test('convertCsvToData function', () => {
  const result = convertCsvToData(csvMock);

  expect(result).toEqual([{
    reference: 123,
    accountNumber: 'NL123',
    description: 'description',
    startBalance: 123,
    mutation: 123,
    endBalance: 123
  }]);
});