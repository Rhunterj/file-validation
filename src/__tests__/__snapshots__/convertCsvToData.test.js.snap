// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertCsvToData function 1`] = `
Array [
  Object {
    "accountNumber": "NL123",
    "description": "description",
    "endBalance": 123,
    "mutation": 123,
    "reference": 123,
    "startBalance": 123,
  },
]
`;
